### PR TITLE
Reduce fuel in new vehicles over time.

### PR DESCRIPTION
**Pull request by Vasyan2006**
#### Summary
```SUMMARY: Balance "New vehicles will be spawn with less fuel, based on the time since the cataclysm"``` 

#### Purpose of change
The fuel will not be infinite resource and will become more valuable. Characters will be forced to find renewable energy source to raid neaby cities.
Because other survivals syphoning fuel too.

#### Describe the solution
~~Quantity of fuel in new vehicles will be decreased based on the time since the start.
It has exponential dependence and after 10 days new cars will have 79% of initial fuel, 50% after 30 days and 12% after 91 days.~~
New vehicles can spawn without fuel and this chanse is increasing over time and based on current NPC settings.
Vehicles without fuel:

| Day  | No NPC | Only static  | Only dynamic (0.1) | Only dynamic (0.2 or above) | Static and dynamic (0.1) |  Static and dynamic (0.2 or above) |
| -----|------- | ------------ | ------------       | ------------                |  ------------            | -------------             |
| 0    | 0%     | 0%           | 0%                 | 0%                          | 0%                       | 0%                        |
| 30   | 0%     | 16%          | 16%                | 30%                         | 30%                      | 41%                       |
| 60   | 0%     | 30%          | 30%                | 50%                         | 50%                      | 65%                       |
| 120  | 0%     | 50%          | 50%                | 75%                         | 75%                      | 87%                       |
| 240  | 0%     | 75%          | 75%                | 99%                         | 99%                      | 99%                       |

This rate can be changed or completely disabled in game settings.
It will only change vehicles which are created in new locations. Already discovered vehicles will not be modified.
It will not change battery charges or vehicles with "undamaged" flag.
It has the same gameplay effect as fuel spoilage, but has another reason and can be tuned more easily.

#### Additional context
Just another kill the fun update. Nothing interesting here.

### CODE REVIEW

**Comment by Kelenius** [comment]
All the more reason to go straight to solars and electric engines. Is that the current goal?

**Comment by Vasyan2006** [comment]
@Kelenius  
Use biofuel, light vehicles and more preparations before expeditions.

**Comment by Kelenius** [comment]
I don't care about this change, I haven't used fuel-based engines in the entire time I've played the game. You don't need to take it personally. I just don't get the purpose of this change. Fuel engines are already inferior, why make them worse?

**Comment by Vasyan2006** [comment]
Fuel engines are not inferior, but electric power it too easy to get. This makes fuel engines useless. But efficiency of solar panels is a good point for another discussion.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -354,6 +354,14 @@ void vehicle::init_state( int init_veh_fuel, int init_veh_status )
         }
     }
 
+    // Reduce fuel in new vehicles over time
+    int fuel_reducing = get_option< int >( "VEHICLE_FUEL_REDUCING" );
+    float fuel_mult_time = 1.0f;
+    if( fuel_reducing > 0 && veh_status != 0 ) {
+        int current_day = to_days< int >( calendar::turn - calendar::time_of_cataclysm );
+        fuel_mult_time = exp( - 0.69f * current_day / fuel_reducing );
```
Comment (mlangsdorf) [inline src/vehicle.cpp]: Please declare the -0.69f as a static_const someplace so it's easy to see and reference.

**Comment by TechyBen** [comment]
Fuel does degrade over time (I've got some here that is only good for patio heater/cooking now and not combustion).

However, as above, I guess for this game we need actual details/stats not just random numbers. The "other survivors" bit does not quite play out, as often you are the lone survivor (other things are not tracked this way, we would need a global loot reduction, not only a fuel one).

But fuel evaporation + deterioration (IIRC Octane percentage?) over time would be fine.

**Comment by Vasyan2006** [comment]
Fuel in new vehicles will be reduced twice every X day, where X is written in game_balance.json. It is set to 30 days as default.

| Day  | Fuel (from initial quantity) |
| ------------- | ------------- |
| 0  | 100%  |
| 10  | 80%  |
| 30  | 50%  |
| 60  | 25%  |
| 364 | 0.02%  |

Initial fuel (100% in the table) is how much fuel is in vehicles right now.

-0.69f was used to change the base of exp() function and it is replaced with pow() function. Hopefully it is more understandable now.

@TechyBen 
Fuel spoilage is more difficult to implement, because it must not be instant change of all gasoline, but slow decreasing of engine power and possible engine malfunctions.

**Comment by Marrim** [comment]
What about fuel in gas stations? They are the main source of fuel, not the broken cars strewn about.

**Comment by Robik81** [comment]
I don't like this change. Vehicles already start with almost empty tanks. Also, I doubt any survivor would stroll around the city and half siphon all cars around. Not to mention in the middle of zombies that left them alone?

> Just another kill the fun update. Nothing interesting here.

Yeah, pretty much.

**Comment by Vasyan2006** [comment]
>What about fuel in gas stations?

I forgot about them. Better to change them later, because item spawn code is much more trickier.

>I doubt any survivor would stroll around the city and half siphon all cars around. 

But this is exactly that my survivors are doing.

**Comment by Robik81** [comment]
> But this is exactly that my survivors are doing.

No, you are not, you siphon them to the last drop, like the rest of us. Also, you have to kill the zombies first.

**Comment by tinukedaya** [comment]
This make no sense as Robik81 said. You don't go and half-suck all the cars. No you go and completely suck half the cars... So if you want to make something like this, change the chance of newly spawned car having any fuel at all. 

Although I'd wait for @mlangsdorf to do his magic in revamping vehicle code, his engine efficiencies etc. before saying there is abundance of fuel.

**Comment by mlangsdorf** [comment]
50% of the fuel in new cars disappearing every game month is an absurdly fast rate. In a month of play, my survivor has barely siphoned 180L of diesel from the vehicles within 150 overmap tiles E/W of my starting location. There's been plenty of vehicles that I've noted as "has lots of diesel" for future refueling, but didn't siphon from because either my vehicle was full or it wasn't convenient to siphon the tanks when I encountered the vehicle.

I agree with tinukedaya that a better approach would be to decrease the chance of a newly spawned vehicle having any vehicle._

**Comment by Maskrow** [comment]
Maybe make another kind of fuel - "bad gasoline" that is used the same way, but damages the engine? And the chance of meeting a car with this "bad" gasoline increasing over time.

**Comment by Vasyan2006** [comment]
- Some cars will be created without fuel, instead of redicung fuel in all new cars,
- A half of cars will be created without fuel at day 60, instead of 30. It is very subjective estimation and depends mostly on personal preferenses. Or tell me how long this time should be, because it is some sort of difficulty settings.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -354,6 +354,14 @@ void vehicle::init_state( int init_veh_fuel, int init_veh_status )
         }
     }
 
+    // Fuel in some vehicles will be syphoned by other survivals.
+    int fuel_reducing = get_option< int >( "VEHICLE_FUEL_REDUCING" );
+    bool has_fuel = 1;
```
Comment (BevapDin) [inline src/vehicle.cpp]: Why are you initializing a `bool` with an `int`?

Comment (Vasyan2006) [reply src/vehicle.cpp]: Isn't it a standart covnersion? This variable will be set to 1 or 0 later, so you suggest to change its type to `int`?

Comment (AMurkin) [reply src/vehicle.cpp]: boolean variables have value true or false.

**Comment by mlangsdorf** [comment]
A multiplicative 50% drop per period produces a very stark curve.  Here's a table for different base percentages:

Period | 100% | 90% | 80% | 70% | 60% | 50%
-- | -- | -- | -- | -- | -- | --
1 | 100.00% | 90.00% | 80.00% | 70.00% | 60.00% | 50.00%
2 | 100.00% | 81.00% | 64.00% | 49.00% | 36.00% | 25.00%
3 | 100.00% | 72.90% | 51.20% | 34.30% | 21.60% | 12.50%
4 | 100.00% | 65.61% | 40.96% | 24.01% | 12.96% | 6.25%
5 | 100.00% | 59.05% | 32.77% | 16.81% | 7.78% | 3.13%
6 | 100.00% | 53.14% | 26.21% | 11.76% | 4.67% | 1.56%
7 | 100.00% | 47.83% | 20.97% | 8.24% | 2.80% | 0.78%
8 | 100.00% | 43.05% | 16.78% | 5.76% | 1.68% | 0.39%
9 | 100.00% | 38.74% | 13.42% | 4.04% | 1.01% | 0.20%
10 | 100.00% | 34.87% | 10.74% | 2.82% | 0.60% | 0.10%
11 | 100.00% | 31.38% | 8.59% | 1.98% | 0.36% | 0.05%
12 | 100.00% | 28.24% | 6.87% | 1.38% | 0.22% | 0.02%

I suggest setting the default to 0.9 - ( NPC spawn factor scaled from 0 to 0.2 ). So if you play a game with the NPC spawn rate at 100, then 50% of new cars don't have gas after two periods, to represent all those NPCs siphoning off vehicles. At NPC spawn rate 0, you're the only person siphoning vehicles and 70% of new vehicles will still have gas.

With the lower rate of decline, setting the period per decline back to 30 days could work.

I welcome alternate analysis and ideas.

**Comment by Vasyan2006** [comment]
Probability to get a car without fuel depends on current NPC settings. Nothing will be changes in a world without NPCs, static NPCs add flat decreasing rate and decreasing from dynamic NPSs depends on their spawn rate. I have not seen any ingame difference between default NPC rate 0.1 and higher values, so fuel stealing by randoms is limited at spawn rate 0.2.

Number of cars without fuel:

| Day  | No NPC | Only static  | Only dynamic (0.1) | Only dynamic (0.2 or above) | Static and dynamic (0.1) |  Static and dynamic (0.2 or above) |
| -----|------- | ------------ | ------------       | ------------                |  ------------            | -------------             |
| 0    | 0%     | 0%           | 0%                 | 0%                          | 0%                       | 0%                        |
| 30   | 0%     | 16%          | 16%                | 30%                         | 30%                      | 41%                       |
| 60   | 0%     | 30%          | 30%                | 50%                         | 50%                      | 65%                       |
| 120  | 0%     | 50%          | 50%                | 75%                         | 75%                      | 87%                       |
| 240  | 0%     | 75%          | 75%                | 99%                         | 99%                      | 99%                       |

All these numbers are very subjective and open for  discussion.

**Comment by mlangsdorf** [comment]
There's a huge difference in the number of NPCs between the default rate of 0.1 and a rate of 7 or 13 or so.

Regardless, recasting your table as % of cars spawning with fuel:

period | static | dynamic minimum | dynamic maximum
-- | -- | -- | --
0 | 100.00% | 100.00% | 100.00%
1 | 70.71% | 50.00% | 35.36%
2 | 50.00% | 25.00% | 12.50%
3 | 35.36% | 12.50% | 4.42%
4 | 25.00% | 6.25% | 1.56%
5 | 17.68% | 3.13% | 0.55%
6 | 12.50% | 1.56% | 0.20%
7 | 8.84% | 0.78% | 0.07%
8 | 6.25% | 0.39% | 0.02%

And we're still at half-life or worse curve with dynamic NPCs enabled.

Instead of pow( 2, -npc_factor * time period ), I'd really prefer pow( 1.11, -factor ) or pow( 1.25, -factor ) for reasons I mentioned above.

**Comment by Vasyan2006** [comment]
But pow( 2, - time ) = pow( 1.11, - 6.9 * time ) = pow( 1.25, - 3.0 * time )
Changing the base will not change the shape of the curve, only scale it over time, so this process will occur faster or slower. Same results can be received by changing the time factor. Look like we are using different names for the same things and it leads to misanderstandings.

**Comment by ZhilkinSerg** [comment]
Marking as Stalled to clean up PR queue. Please ping me to get it reopened when you feel you can continue working on it.

**Comment by Vasyan2006** [comment]
Exponential functions e^(-time/T) are widely used in simulation of different physical processes and they have one and the only one coefficient T (or halt-time or parameter or characteristic time or whatever you call it).
e = 2.71... is one of mathematical magic numbers but it is not required unless you are working with complex numbers or doing other stupid things.
You can switch from e^(-time/T) to 2^(-time/T1) or 1.1^(-time/T2) by changing T to T1 or T2 and ingame results will be the same.
Here are plots:

untitled 2


By default e^(-time/30), 2^(-time/30) and 1.1^(-time/30) are different, but after changing coeffitient T they will become the same.
See e^(-0.65 time/30), 2^(-time/30) and 1.1^(-7.4 time/30) as example. Plots are very similar because this transformation is just moving a multiplier from one place to another.
I intentionall used slightly wrong multipliers (0.65 instead of ln2 = 0.69... and 7.4 instead of ln2 / ln1.1 = 7.27...) to make these plots visible, otherwise they will overlap.  

And also same results for e^(-0.65 time/60), 2^(-time/60) and 1.1^(-7.4 time/60).

You can change the base from 2 to 1.001 and still have the same ingame results (but in this case time step will be larger then characteristic time, so dont expect any presicion).

Very good example of this is a monster evolution. Instead of using exp function and get probability for any time interval it is overcomplicated, entire time divided into periods, for every period it calculates mutation chance and mutation time or time for the next chech and the loop starts again.
It is just using brute force instead of elegant solutions.
Same is for calculation of cold/hot food statuses - just a lot of unnesessary variables and their recalculations.

I used to work on simulations of different physical processes and all these transformations are obvious for me. A lot of mathematical tools were created in last centuries and they can be very handy if used correctly. But fine, lets keep this PR closed just because it is difficult to understand.

**Comment by ZhilkinSerg** [comment]
> But fine, lets keep this PR closed just because it is difficult to understand.

That is invalid statement.

**Pull request closed without merging**